File: 2685-first-completely-painted-row-or-column/first-completely-painted-row-or-column.py
class Solution:
    def firstCompleteIndex(self, a: List[int], mat: List[List[int]]) -> int:
        n, m = len(mat), len(mat[0])
        R, C = [m] * n, [n] * m
        idx = [None] * (n * m + 1)
        for i in range(n):
            for j in range(m):
                idx[mat[i][j]] = (i, j)
        
        for i in range(len(a)):
            x, y = idx[a[i]]
            R[x] -= 1
            C[y] -= 1
            if R[x] == 0 or C[y] == 0:
                return i 
        
        return -1 

File: 2685-first-completely-painted-row-or-column/first-completely-painted-row-or-column.java
class Solution {
    public int firstCompleteIndex(int[] a, int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        int[] C = new int[m];
        int[] R = new int[n];
        int[] idxR = new int[n * m + 1] ;
        int[] idxC = new int[n * m + 1] ;
        for(int i = 0 ; i < n ; i ++){
            for(int j = 0 ; j < m ; j ++){
                idxR[mat[i][j]] = i ;
                idxC[mat[i][j]] = j ;
            }
        }  
        for(int i = 0 ; i < n * m ; i ++){
            R[idxR[a[i]]] ++ ;
            C[idxC[a[i]]] ++ ;
            if(R[idxR[a[i]]] == m || C[idxC[a[i]]] == n){
                return i ;
            }
        }
        return -1 ;
    }
}